### PR TITLE
Virtual Scrolling for Resource List Views

### PR DESCRIPTION
**Pull request by amrutac**
### Description

Improves performance for huge resource lists (they don't crash) using `react-virtualized`.

### Testing

1. Run using load testing ServiceWorker: `./bin/bridge --load-test-factor 30`
2. View "Workloads"->"Pods"
3. Observe the smooth scrolling

Addresses https://jira.coreos.com/browse/CONSOLE-467

### CODE REVIEW

File: frontend/public/components/factory/list.jsx [hunk 1]
```diff
@@ -191,9 +193,64 @@ export const WorkloadListHeader = props => <ListHeader>
...
+    return (
+      <CellMeasurer
+        cache={this.measurementCache}
+        columnIndex={0}
+        key={key}
+        rowIndex={index}
+        parent={parent}>
+        {({measure}) => (
+          <div style={style}>
+            <Row key={key} obj={data[index]} expand={expand} kindObj={kindObj} index={index} onLoad={measure} />
```
[the remark this answers is not in the document]
Comment (amrutac) [reply frontend/public/components/factory/list.jsx]: Yeah, removed it, had added it while I was playing around with the props and trying to get the cell measurer to work.

File: frontend/public/components/factory/list.jsx [hunk 2]
```diff
@@ -200,9 +202,62 @@ export const WorkloadListHeader = props => <ListHeader>
...
 </ListHeader>;
 
-const Rows = ({data, expand, Row, kindObj}) => <div className="co-m-table-grid__body">
-  {data.map(obj => <Row key={obj.rowKey || obj.metadata.uid} obj={obj} expand={expand} kindObj={kindObj} />)}
-</div>;
+export class Rows extends React.Component {
+  constructor(props) {
+    super(props);
+
+    this.measurementCache = new CellMeasurerCache({
```
Comment (kans) [inline frontend/public/components/factory/list.jsx]: `CellMeasurerCache` caches by default by row and column but it should probably be using [keyMapper](https://github.com/bvaughn/react-virtualized/blob/master/docs/CellMeasurer.md#prop-types-1) and metadata.uid

**Comment by amrutac** [comment]
Tests are failing because of https://github.com/bvaughn/react-virtualized/issues/876. Looking

**Comment by amrutac** [comment]
> The Rows aren't using the styles because of the wrapper div.
Yes, I added the wrapper div to row to mitigate the choppiness. But the row heights seem fine to me, is there any particular resource list that you are looking at?

> Othewise, scrolling Pods feels choppy. I wonder if its because of instead of giving Rows a fixed height.
Yeah, that's what suspected. But wrapper div seems to fix it.

File: frontend/public/components/factory/list.jsx [hunk 3]
```diff
@@ -241,7 +240,6 @@ export class Rows extends React.Component {
             {({width}) => <div ref={registerChild}>
               <VirtualList
                 autoHeight
-                data={data}
```
Comment (kans) [inline frontend/public/components/factory/list.jsx]: I think we need to pass data to tell the list to rerender when data changes, but there may be a better way.

File: frontend/public/style/_overrides.scss [hunk 4]
```diff
@@ -189,3 +189,11 @@ tags-input .autocomplete .suggestion-item em {
 .modal-footer .alert {
   text-align: left;
 }
+
+// FIXME: Pass as `style` prop to `List` once this is resolved (https://github.com/bvaughn/react-virtualized/issues/876)
+.ReactVirtualized__Grid.ReactVirtualized__List {
```
Comment (alecmerdler) [inline frontend/public/style/_overrides.scss]: @kans Is this an acceptable solution for now? I played around with `ReactDOM.createPortal()`, but it requires polluting our `props` with X and Y positioning (see https://codepen.io/jscottsmith/pen/xYaawK). We can open a PR upstream that lets us override it in the `style` prop passed to `List` if needed.

**Comment by kans** [comment]
I played with this PR some.  Scrolling is still janky on `Pods`, even with a fixed height.  Profiling showed that we can't mount PodRows fast enough to keep up with scrolling.  On my machine, mounting takes anywhere from 2-4 ms per Row (20 rows visible).  Of that time, we spend ~1.5 ms on Labels alone, and ~.4-.7 on Cogs.  Links are surprisingly expensive at .3ms.  Cogs and Links are both `connected` to redux.  `ResourceLink`s spend 3/5 of their time in `ResourceIcon`s (connected).  

Removing Cogs, Labels, and increasing the fixed height to 75px still results in dropped frames when I  scroll.  

Proposals:
1. We should consider dropping labels or only optionally showing them on some user action (like for Nodes). 
2. `ResourceCog` could probably have a much later binding to `kindObj` because we don't care about the `kindObj` until the user clicks on the cog - they are all rendered the same.
3. `ResourceIcon` binds to redux to look up `kinds` and then injects a `className` (`co-m-resource-${kindStr.toLowerCase()}`), which is *alot* of work to turn "Node" into "node".  I have some fuzzy recollection this was done to support CRDs for ALM?  Maybe we could memoize something here or use a late binding when we need it.

We should address the performance issues of `ResourceCog` and `ResourceIcon` in other PRs.

**Comment by alecmerdler** [comment]
Closed in favor of https://github.com/openshift/console/pull/92

**Pull request closed without merging**